### PR TITLE
Add `Optional` Environment

### LINKED ISSUE #23146: [Investment] Python Matrix on optional dependency
@lmazuel commented on [Wed Nov 18 2020](https://github.com/Azure/azure-sdk-pr/issues/656)

Python is designed to have a lot of optional dependencies, meaning the behavior of the package can automatically change if a package is detected to be installed on the fly (no compilation needed here).

The big items here are:
- Is our async HTTP lib installed (aiohttp)? What impact on code stability, and perf?
- Is tracing lib installed (opentelemetry)? What impact on code stability, and perf?

Right now, our CI system is testing WITH aiohttp installed (all the time), and WITHOUT tracing installed (all the time). We have no proof that our packages are acting correctly if optional dependencies are installed (or not). For instance, the common scenario in Python is "no async / no tracing", but we actually just test that combination on Python 2 (which is not on purpose, just there is no async on Py2...)

Example of scenario that could happen, and that this work tries to avoid:
- Sdk XYZ with no aiohttp installed (no CI test that), the import could breaks if we messed up the optional part
- EventHub with opentelemetry installed, will automatically start to trace the code, that could break (we should NEVER break if tracing creates an error, but we don't know).

This issue is about to have CI dedicated to the combination stdlib/aiohttp/tracing and have a nice way to test that and mesure perf and impact. This is part engsys work, but I expect some test framework to make it available.

Estimate in weeks (one sprint)

### PR DESCRIPTION
**Pull request by scbedd**
# Optional Requirements Environment

Resolves #23146 

This PR adds an additional `tox` environment named `optional`. This environment is designed to work with a new config key in `pyproject.toml`. An `optional` environment is designed to allow extremely easy access to various scenarios that a dev wants to run as part of their nightly job. Let's look at an example with `azure-core` as the subject.

- [Link to Source](https://github.com/Azure/azure-sdk-for-python/blob/feature/add-optional-env/sdk/core/azure-core/pyproject.toml#L11)

```toml
[tool.azure-sdk-build]
mypy = true
type_check_samples = true
verifytypes = true
pyright = false

# here are the new entries

[[tool.azure-sdk-build.optional]]
name = "no_requests"
install = []
uninstall = ["requests"]
additional_pytest_args = ["-k", "*_async.py"]

[[tool.azure-sdk-build.optional]]
name = "no_aiohttp"
install = []
uninstall = ["aiohttp"]
additional_pytest_args = ["-k", "not *_async.py"]
```

Note the two additional `[[tool.azure-sdk-build.optional]]` items. A user will add one or multiple of these configurations as they need! They are fairly self explanatory, but for the sake of completeness.

|Name|Purpose|
|---|---|
| `name` | The name of the environment when logging. A user can pass `tox -e optional -c ../../../eng/tox/tox.ini --root . -- --optional=<name>` if they wish to invoke a single environment.
| `install` | An array of valid package specifiers. These are **added** _after_ your package + dev_requirements are installed, but _before_ the `uninstall` config item is honored. |
| `uninstall` | An array of valid package specifiers. These are **removed** _after_ `package`, `dev_requirements`, and `install` have been installed.|
|`additional_pytest_args` | An array of valid pytest arguments. These will be appended to the pytest invocation. Useful to exclude test files that you know won't work, like excluding `not *_async.py` when `aiohttp` isn't installed. |

By default, the new tox environment will iterate through the `optional` environments, invoking them one at a time. These checks will be added to the `nightly` build.

Passing an `optional` or `-o` argument through to the tox environment will allow you to run specific optional configurations.

```bash
tox -e optional -c ../../../eng/tox/tox.ini --root . -- --optional no_requests
```

### CODE REVIEW

File: tools/azure-sdk-tools/ci_tools/scenario/generation.py [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+        f for f in os.listdir(dist_dir) if f.endswith(".whl" if package_type == "wheel" else ".tar.gz")
+    ]
+
+    if not in_ci():
+        logging.info("Cleaning up build directories and files")
+        cleanup_build_artifacts(target_setup)
+    return prebuilt_packages
+
+
+def prepare_and_test(mapped_args: argparse.Namespace) -> int:
```
Comment (mccoyp) [inline tools/azure-sdk-tools/ci_tools/scenario/generation.py:387]: Nit: if this only handles optional environments, we'll probably want to make the function name more indicative of that. Maybe `prepare_and_test_optional`?

**Review by mccoyp — COMMENTED** [review]
> I see some value in being able to invoke the run_optional script to simply prepare your current venv.

Is there a use case you have in mind? Using a unique venv per environment sounds like a safer option, but I don't have a strong opinion

File: eng/tox/create_package_and_install.py [hunk 2]
```diff
@@ -148,117 +82,7 @@ def build_and_discover_package(setuppy_path, dist_dir, target_setup, package_typ
...
-                                    env=dict(os.environ, PIP_EXTRA_INDEX_URL=""),
-                                )
-                            except CalledProcessError as e:
-                                req_name, req_specifier = parse_require(addition)
-                                non_present_reqs.append(req_name)
-
-                        additional_downloaded_reqs = [
-                            os.path.abspath(os.path.join(tmp_dl_folder, pth)) for pth in os.listdir(tmp_dl_folder)
-                        ] + [get_package_from_repo(relative_req).folder for relative_req in non_present_reqs]
+    create_package_and_install(args.distribution_directory, args.target_setup, args.skip_install, args.cache_dir, args.work_dir, args.force_create, args.package_type, args.pre_download_disabled)
```
[the remark this answers is not in the document]
Comment (scbedd) [reply eng/tox/create_package_and_install.py:85]: I totally agree.

File: eng/tox/run_optional.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        "--optional",
+        dest="optional",
+        help="The target environment. If not matched to any of the named optional environments, hard exit. If not provided, all optional environments will be run.",
+        required=False,
+    )
+
+    parser.add_argument(
+        "--temp",
+        dest="temp_dir",
+        help="The target environment. If not matched to any of the named optional environments, hard exit. If not provided, all optional environments will be run.",
```
[the remark this answers is not in the document]
Comment (scbedd) [reply eng/tox/run_optional.py:23]: Yeah, and it's used that way, I just need to fix the `help` text 😂

File: tools/azure-sdk-tools/ci_tools/scenario/generation.py (hunk 1 above)
Comment (scbedd) [reply tools/azure-sdk-tools/ci_tools/scenario/generation.py:387]: Agreed.

**Comment by scbedd** [comment]
> > I see some value in being able to invoke the run_optional script to simply prepare your current venv.
> 
> Is there a use case you have in mind? Using a unique venv per environment sounds like a safer option, but I don't have a strong opinion

Yeah. I'm just gonna do that. It definitely is safer.

**Comment by scbedd** [comment]
Closing to work through merge conflicts and failing CI checks.

**Comment by scbedd** [comment]
@mccoyp one may notice that this PR doesn't actually add `optional` to the `PR` or `nightly` checks. This is **intentional**. I want to get the bulk of the codemovement in, then I will submit a follow-up PR that:

- Actually enables the `optional` environment for the nightly / CI build (not PR)
- fixes the various import errors that exist if `requests` isn't installed when attempting to invoke the tests.
  - These could be test, could be actual issues in `azure-core`. Let's get the PR smaller so it's easier to reason through if there are actual changes to `azure-core` necessary.

File: tools/azure-sdk-tools/ci_tools/build.py [hunk 4]
```diff
@@ -126,6 +126,20 @@ def build() -> None:
     )
 
 
+def cleanup_build_artifacts(build_folder):
+    # clean up egginfo
+    results = glob.glob(os.path.join(build_folder, "*.egg-info"))
+
+    if results:
+        print(results[0])
```
Comment (mccoyp) [inline tools/azure-sdk-tools/ci_tools/build.py:134]: Leftover from debugging?

**Review by mccoyp — APPROVED** [review]
This is really awesome! Agreed that it'll be best to get these changes in first before getting things enabled

**Pull request merged**